modules/api/pkg/ee/clusterbackup/backupstoragelocation/handler.go

```diff
@@ -0,0 +1,458 @@
+	bslFullName := fmt.Sprintf("%s-%s-%s", bslName, hash, clusterID)
+	existingSecret := true
+	secret := &corev1.Secret{}
+	if err := client.Get(ctx, types.NamespacedName{Name: secretRefName, Namespace: clusterbackup.UserClusterBackupNamespace}, secret); err != nil {
+		log.Logger.Errorf("secret doesn't exist for ref: %s in the cluster, creating new secret", secretRefName)
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		if apierrors.IsNotFound(err) || secret.Name == "" {
+			existingSecret = false
+			dataBytes, err := backupProvider.GetStorageLocationCreds(ctx, secretRefName)
+			if err != nil {
+				return nil, fmt.Errorf("please provide correct creds, error: %v", err.Error())
+			}
+			if dataBytes["accessKeyId"] == nil || dataBytes["secretAccessKey"] == nil {
+				return nil, fmt.Errorf("bsl credentials secret is not set correctly")
+			}
+			cloudCreds, err := getVeleroCloudCredentials(dataBytes["accessKeyId"], dataBytes["secretAccessKey"])
+			if err != nil {
+				return nil, fmt.Errorf("failed to generate Velero cloud-credentials template: %w", err)
+			}
+			if bsl.Spec.Credential.Key == "" {
+				bsl.Spec.Credential.Key = defaultCloudCredentialsSecretKeyName
+			}
+			secret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretRefName,
+					Namespace: clusterbackup.UserClusterBackupNamespace,
+					Labels:    getBSLLabels(bslFullName, projectID, clusterID),
+				},
+				Data: map[string][]byte{
+					bsl.Spec.Credential.Key: cloudCreds,
+				},
+			}
+		}
+	}
+
+	bsl.ObjectMeta = metav1.ObjectMeta{
+		Name:      bslFullName,
+		Namespace: clusterbackup.UserClusterBackupNamespace,
+		Labels:    getBSLLabels(bslName, projectID, clusterID),
+	}
+
+	if !existingSecret {
+		bsl.Spec.Credential = &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: secret.Name,
+			},
+			Key: credentialsSecretKeyName,
+		}
+	}
+
+	if err := client.Create(ctx, bsl); err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	ownerReferences := metav1.OwnerReference{
+		APIVersion: velerov1.SchemeGroupVersion.String(),
+		Kind:       bslKind,
+		Name:       bslFullName,
+		UID:        bsl.UID,
+	}
+	if !existingSecret {
+		secret.ObjectMeta.OwnerReferences = append(secret.ObjectMeta.OwnerReferences, ownerReferences)
+		if err := client.Create(ctx, secret); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+	} else {
+		ownerRefExist := false
+		for _, ownerRef := range secret.ObjectMeta.OwnerReferences {
+			if ownerRef.UID == bsl.UID {
+				ownerRefExist = true
+				break
+			}
+		}
+		if !ownerRefExist {
+			updatedSecret := secret.DeepCopy()
+			updatedSecret.ObjectMeta.OwnerReferences = append(updatedSecret.ObjectMeta.OwnerReferences, ownerReferences)
+			if err := client.Patch(ctx, updatedSecret, ctrlruntimeclient.MergeFrom(secret)); err != nil {
+				return nil, common.KubernetesErrorToHTTPError(err)
+			}
+		}
+	}
```

Comment:
We should only add owner references to the secrets that we create against the new BSLs no? I've not visualized this code in my mind but if this is adding owner reference to a secret that's already being used somewhere then deleting the new BSL would delete the existing secret that's being used by velero for some other BSL.